Refactor VolumeHost interface to be kubelet specific

jsafrane opened the pull request:
All volume plugins need a VolumeHost to interact with the host. Most of its functionality is needed only by kubelet, so let's make the interface kubelet-only and split 'cloud' and 'kubeClient' (needed by all non-kubelet functionality) as part of Init() method.

edit: As result, we can remove VolumeHost implementation from PV controller and attach/detach controller (see the second commit, "Remove VolumeHost implementation from controllers") and in future we can easily use volume plugins in admission plugins (e.g. to check StorageClass parameters or so).

Partly fixes #14217

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/39425)
<!-- Reviewable:end -->

jsafrane commented:
I added some checks to report error when something goes wrong + I pruned volume interfaces a bit to get rid of VolumeHost in Recyclers and Deleters.

jsafrane commented:
Ouch, 100 changed files. The changes in individual commits are quite simple, however we have many volume plugins and I had to change all of them.

rootfs commented on pkg/volume/plugins.go:
```diff
@@ -183,9 +183,6 @@ type VolumeHost interface {
 	// directory might not actually exist on disk yet.
 	GetPodPluginDir(podUID types.UID, pluginName string) string
 
-	// GetKubeClient returns a client interface
```
instead of directly exposing client and cloud provider and refactoring all plugins' `Init()`, would you add a constructor to initialize a `VolumeHost` with `client` and `cloudprovider`? Then call this constructor in `InitPlugins` and get a `VolumeHost`, then we should be able to reuse `plugin.Init(host)`

jsafrane replied:
Whole point of this PR is to get rid of VolumeHost implementation from controllers. I don't want to create a dummy VolumeHost with fake methods like `GetPodPluginDir`.

rootfs commented:
This is a nice refactoring. It removes a lots of unwanted code in controllers.

jsafrane commented:
@k8s-bot unit test this issue: #39471

jsafrane committed:
Remove VolumeHost implementation from controllers

jsafrane committed:
Add VolumeHost checks where necessary.

jsafrane committed:
Prune volume.Deleter and .Recycler interfaces.

Both Deleter and Recycler inherit GetPath() method from volume.Volume
interface without any reason - the code that recycles/deletes PVs does
not need GetPath (or any method from Volume interface), so let's remove
it.

jingxu97 commented:
I am still not very clear about our VolumeHost implementation. There are two volume_host.go, one is under /kubulet, the other is under /controller/volume/persistentvolume. It seems like this change is for the second one. How about the first one? Is there any change we should make?

jsafrane commented:
There are at least three implementations, one in kubelet, one in PV controller and one in attach/detach controller (+ several mock implementations for tests). Goal of this PR is to remove all of them except the kubelet one.

jingxu97 commented:
@saad-ali, could you help review this one?

jsafrane commented:
Rebased.

jsafrane commented:
@pmorie, @saad-ali, this is tedious to rebase, can you please review this PR?

pmorie commented:
@jsafrane Really sorry for the latency on this one.  I'll review it now.

pmorie commented on pkg/controller/volume/persistentvolume/pv_controller_base.go:
```diff
@@ -90,7 +90,7 @@ func NewController(p ControllerParameters) (*PersistentVolumeController, error)
 	}
 
-	if err := controller.volumePluginMgr.InitPlugins(p.VolumePlugins, controller); err != nil {
+	if err := controller.volumePluginMgr.InitPlugins(p.VolumePlugins, nil, p.KubeClient, p.Cloud); err != nil {
```
Can you add a comment after `nil`, like so:

```go
InitPlugins(p.VolumePlugins, nil /* volumeHost */, p.KubeClient, p.Cloud)
```

jsafrane replied:
Fixed

pmorie reviewed (commented):
LGTM, one nit

pmorie commented:
@jsafrane this is good medicine -- thanks a lot for the patch

jsafrane committed:
Refactor VolumeHost to be kubelet specific.

All volume plugins need a VolumeHost to interact with the host. Most of its
functionality is needed only by kubelet, so let's make the interface
kubelet-only and split 'cloud' and 'kubeClient' (needed by all non-kubelet
functionality) as arguments of Init() method.

jsafrane commented:
@thockin, PTAL and approve
(this should be faster than asking 3 different for approval...)

pmorie commented:
/lgtm

thockin commented:
I'm torn on this.  The point of VolumeHost was to offer services to plugins without having to update the signature of Init() every time one plugin needs something new, and to have a rigid interface against which we could test.  This PR does the opposite of the first goal - see the need to pass `clientset` to Init() even though almost nobody needs it.  Additionally, once you make the `host` arg nil-able, you have to add tests for whether it is nil (which you did) which are bound to be forgotten.

I feel like this is making the plugins overall more brittle and more coupled.  Isn't there something else we can do?  This implies that some methods can depend on host being provided and some methods can not.  Can we fracture on that edge instead?  Can we at the VERY least document which sub-interfaces and methods can rely on having a host present?

pmorie commented:
@thockin I think that this moves us in the right direction, but I do see your point.  There are basically two classes of dependencies - those that are about the host where something is mounted or unmounted, and those that are for things like provisioning, which doesn't run on a node.  I think the split is that mount/unmount/attach/detach need volume host, and everything else doesn't.  I would be cool introducing a different object to represent the non-host ops, but I have considered volumehost to be poluted for quite some time.  How you would feel about merging this, and then introducing another dependency group in a follow up?  (assuming that we add clarifications to the interface declarations about what is injected with what)

jsafrane commented:
I'm closing this PR as I am implementing mount pods and both controllers and kubelet will need additional method(s) in VolumeHost.
